diff -ruN jsoncpp-1.7.2/CMakeLists.txt jsoncpp-1.7.2.no-abi/CMakeLists.txt
--- jsoncpp-1.7.2/CMakeLists.txt	2016-03-25 21:09:15.000000000 +0100
+++ jsoncpp-1.7.2.no-abi/CMakeLists.txt	2016-07-19 09:52:51.671872314 +0200
@@ -47,12 +47,13 @@
 # version parts in ${OUPUT_PREFIX}_MAJOR, _MINOR, _PATCH.
 # If parse succeeds then ${OUPUT_PREFIX}_FOUND is TRUE.
 MACRO(jsoncpp_parse_version VERSION_TEXT OUPUT_PREFIX)
-    SET(VERSION_REGEX "[0-9]+\\.[0-9]+\\.[0-9]+(-[a-zA-Z0-9_]+)?")
+    SET(VERSION_REGEX "([0-9]+)\\.([0-9]+)\\.([0-9]+)([.-]?[a-zA-Z0-9][-a-zA-Z0-9_]*)?")
     IF( ${VERSION_TEXT} MATCHES ${VERSION_REGEX} )
-        STRING(REGEX MATCHALL "[0-9]+|-([A-Za-z0-9_]+)" VERSION_PARTS ${VERSION_TEXT})
+        STRING(REGEX MATCHALL "[0-9]+|([A-Za-z0-9][-A-Za-z0-9_]*)" VERSION_PARTS ${VERSION_TEXT})
         LIST(GET VERSION_PARTS 0 ${OUPUT_PREFIX}_MAJOR)
         LIST(GET VERSION_PARTS 1 ${OUPUT_PREFIX}_MINOR)
         LIST(GET VERSION_PARTS 2 ${OUPUT_PREFIX}_PATCH)
+        # ignore "additional" extensions
         set_using_dynamic_name( "${OUPUT_PREFIX}_FOUND" TRUE )
     ELSE( ${VERSION_TEXT} MATCHES ${VERSION_REGEX} )
         set_using_dynamic_name( "${OUPUT_PREFIX}_FOUND" FALSE )
@@ -64,14 +65,21 @@
 #SET( JSONCPP_VERSION_MAJOR X )
 #SET( JSONCPP_VERSION_MINOR Y )
 #SET( JSONCPP_VERSION_PATCH Z )
-SET( JSONCPP_VERSION 1.7.2 )
+IF ( NOT JSONCPP_VERSION )
+    SET( JSONCPP_VERSION 1.7.2 )
+ENDIF()
 jsoncpp_parse_version( ${JSONCPP_VERSION} JSONCPP_VERSION )
-#IF(NOT JSONCPP_VERSION_FOUND)
-#    MESSAGE(FATAL_ERROR "Failed to parse version string properly. Expect X.Y.Z")
-#ENDIF(NOT JSONCPP_VERSION_FOUND)
+IF(NOT JSONCPP_VERSION_FOUND)
+    MESSAGE(FATAL_ERROR "Failed to parse version string properly. Expect X.Y.Z")
+ENDIF(NOT JSONCPP_VERSION_FOUND)
+IF ( NOT JSONCPP_SOVERSION )
+    SET( JSONCPP_SOVERSION ${JSONCPP_VERSION_MAJOR})
+ENDIF()
+MESSAGE(STATUS "JsonCpp Version: ${JSONCPP_VERSION}, .so version ${JSONCPP_SOVERSION}")
+
+
 SET( JSONCPP_USE_SECURE_MEMORY "0" CACHE STRING "-D...=1 to use memory-wiping allocator for STL" )

-MESSAGE(STATUS "JsonCpp Version: ${JSONCPP_VERSION_MAJOR}.${JSONCPP_VERSION_MINOR}.${JSONCPP_VERSION_PATCH}")
 # File version.h is only regenerated on CMake configure step
 CONFIGURE_FILE( "${PROJECT_SOURCE_DIR}/src/lib_json/version.h.in"
                 "${PROJECT_SOURCE_DIR}/include/json/version.h"
diff -ruN jsoncpp-1.7.2/src/lib_json/CMakeLists.txt jsoncpp-1.7.2.no-abi/src/lib_json/CMakeLists.txt
--- jsoncpp-1.7.2/src/lib_json/CMakeLists.txt	2016-03-25 21:09:15.000000000 +0100
+++ jsoncpp-1.7.2.no-abi/src/lib_json/CMakeLists.txt	2016-07-19 09:46:52.758536736 +0200
@@ -42,7 +42,7 @@
 IF(BUILD_SHARED_LIBS)
     ADD_DEFINITIONS( -DJSON_DLL_BUILD )
     ADD_LIBRARY(jsoncpp_lib SHARED ${PUBLIC_HEADERS} ${jsoncpp_sources})
-    SET_TARGET_PROPERTIES( jsoncpp_lib PROPERTIES VERSION ${JSONCPP_VERSION} SOVERSION ${JSONCPP_VERSION_MAJOR})
+    SET_TARGET_PROPERTIES( jsoncpp_lib PROPERTIES VERSION ${JSONCPP_VERSION} SOVERSION ${JSONCPP_SOVERSION})
     SET_TARGET_PROPERTIES( jsoncpp_lib PROPERTIES OUTPUT_NAME jsoncpp
                            DEBUG_OUTPUT_NAME jsoncpp${DEBUG_LIBNAME_SUFFIX} )

@@ -61,7 +61,6 @@

 IF(BUILD_STATIC_LIBS)
     ADD_LIBRARY(jsoncpp_lib_static STATIC ${PUBLIC_HEADERS} ${jsoncpp_sources})
-    SET_TARGET_PROPERTIES( jsoncpp_lib_static PROPERTIES VERSION ${JSONCPP_VERSION} SOVERSION ${JSONCPP_VERSION_MAJOR})
     SET_TARGET_PROPERTIES( jsoncpp_lib_static PROPERTIES OUTPUT_NAME jsoncpp
                            DEBUG_OUTPUT_NAME jsoncpp${DEBUG_LIBNAME_SUFFIX} )

