diff -Naur safestring/include/safe_lib.h safestring-0.0.2/include/safe_lib.h
--- safestring/include/safe_lib.h	2018-01-26 01:48:48.000000000 +0100
+++ safestring-0.0.2/include/safe_lib.h	2018-01-26 16:55:07.875693637 +0100
@@ -33,6 +33,7 @@
 #ifndef __SAFE_LIB_H__
 #define __SAFE_LIB_H__
 
+#include <wchar.h>
 #include "safe_types.h"
 #include "safe_lib_errno.h"
 
diff -Naur safestring/include/safe_lib.hpp safestring-0.0.2/include/safe_lib.hpp
--- safestring/include/safe_lib.hpp	1970-01-01 01:00:00.000000000 +0100
+++ safestring-0.0.2/include/safe_lib.hpp	2018-01-26 15:52:39.693696282 +0100
@@ -0,0 +1,35 @@
+/*!
+ * @section LICENSE
+ *
+ * @copyright
+ * Copyright (c) 2015-2017 Intel Corporation
+ *
+ * @copyright
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * @copyright
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * @copyright
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * @section DESCRIPTION
+ *
+ * @file safe_lib.hpp
+ *
+ * @brief C++ wrapper for Safe String Library
+ */
+#ifndef SAFELIB_HPP
+#define SAFELIB_HPP
+
+extern "C" {
+#include "safe_lib.h"
+}
+
+#endif /* SAFELIB_HPP */
diff -Naur safestring/include/safe_str_lib.h safestring-0.0.2/include/safe_str_lib.h
--- safestring/include/safe_str_lib.h	2018-01-26 01:48:48.000000000 +0100
+++ safestring-0.0.2/include/safe_str_lib.h	2018-01-26 16:48:09.642318573 +0100
@@ -32,6 +32,8 @@
 #ifndef __SAFE_STR_LIB_H__
 #define __SAFE_STR_LIB_H__
 
+#include <ctype.h>
+
 #include "safe_lib.h"
 
 /*
diff -Naur safestring/safeclib/abort_handler_s.c safestring-0.0.2/safeclib/abort_handler_s.c
--- safestring/safeclib/abort_handler_s.c	2018-01-26 01:48:48.000000000 +0100
+++ safestring-0.0.2/safeclib/abort_handler_s.c	2018-01-26 15:59:40.506068243 +0100
@@ -29,6 +29,8 @@
  *------------------------------------------------------------------
  */
 
+#include <stdlib.h>
+
 #include "safeclib_private.h"
 
 /**
diff -Naur safestring/safeclib/strcasecmp_s.c safestring-0.0.2/safeclib/strcasecmp_s.c
--- safestring/safeclib/strcasecmp_s.c	2018-01-26 01:48:48.000000000 +0100
+++ safestring-0.0.2/safeclib/strcasecmp_s.c	2018-01-26 16:00:21.448221131 +0100
@@ -29,6 +29,8 @@
  *------------------------------------------------------------------
  */
 
+#include <ctype.h>
+
 #include "safeclib_private.h"
 #include "safe_str_constraint.h"
 #include "safe_str_lib.h"
diff -Naur safestring/safeclib/strcasestr_s.c safestring-0.0.2/safeclib/strcasestr_s.c
--- safestring/safeclib/strcasestr_s.c	2018-01-26 01:48:48.000000000 +0100
+++ safestring-0.0.2/safeclib/strcasestr_s.c	2018-01-26 16:01:06.648285921 +0100
@@ -29,6 +29,8 @@
  *------------------------------------------------------------------
  */
 
+#include <ctype.h>
+
 #include "safeclib_private.h"
 #include "safe_str_constraint.h"
 #include "safe_str_lib.h"
